commcare_connect/opportunity/utils/completed_work.py

```diff
@@ -276,3 +277,63 @@ def update_work_payment_date(access: OpportunityAccess):
 
     if works_to_update:
         CompletedWork.objects.bulk_update(works_to_update, ["payment_date"])
+
+
+def get_uninvoiced_completed_works_qs(opportunity, start_date=None, end_date=None):
+    query = CompletedWork.objects.filter(
+        opportunity_access__opportunity=opportunity,
+        status=CompletedWorkStatus.approved,
+        invoice__isnull=True,
+    )
+
+    if start_date:
+        query = query.filter(status_modified_date__date__gte=start_date)
+    if end_date:
+        query = query.filter(status_modified_date__date__lte=end_date)
+
+    return query
+
+
+def get_uninvoiced_visit_items(opportunity, start_date=None, end_date=None, limit=None):
+    from commcare_connect.opportunity.visit_import import get_exchange_rate
+
+    completed_works_qs = get_uninvoiced_completed_works_qs(opportunity, start_date, end_date)
+
+    monthly_pu_records = (
+        completed_works_qs.annotate(
+            month_approved=TruncMonth("status_modified_date"),
+        )
+        .values("payment_unit", "month_approved")
+        .annotate(
+            payment_unit_name=F("payment_unit__name"),
+            payment_unit_amount=F("payment_unit__amount"),
+            record_count=Count("id"),
+            currency=F("opportunity_access__opportunity__currency"),
+            total_amount_usd=Sum("saved_payment_accrued_usd"),
+            total_amount_local=Sum("saved_payment_accrued"),
+        )
+    )
+
+    limited_pu_records = monthly_pu_records[:limit] if limit else monthly_pu_records
+    invoice_items = []
+
+    for record in limited_pu_records:
+        exchange_rate = get_exchange_rate(opportunity.currency, record["month_approved"])
```

Comment:
this should probably be done once per month and cached, not once for every record in the queryset (up to  thousands), especially since there are only expected to be one or two unique values.

Reply:
See https://github.com/dimagi/commcare-connect/pull/839/commits/897f0d93939415de7b2c425ca7e8792eee6ae940 and https://github.com/dimagi/commcare-connect/pull/839/commits/c84d86da456328705d4dcc1abcf7844100cfc78f (second commit considers currency + month)